### PR TITLE
TDL-17695 Add custom error handling

### PR DESCRIPTION
**Pull request by namrata270998**
# Description of change
- Added the error handling for error codes mentioned in the [doc link1](https://developers.google.com/analytics/devguides/reporting/core/v4/errors) and [doc link2](https://developers.google.com/analytics/devguides/config/mgmt/v3/errors)
- Added unit tests to ensure that the code works expectedly,

# QA steps
 - For the corresponding error codes, the tap should throw errors with appropriate messages.
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_google_analytics/client.py [hunk 1]
```diff
@@ -255,12 +325,8 @@ def _make_request(self, method, url, params=None, data=None):
             response = self.session.post(url, headers=headers, params=params, json=data, timeout=self.request_timeout)
         else:
             response = self.session.request(method, url, headers=headers, params=params, timeout=self.request_timeout)
-
-        error_message = _is_json(response) and response.json().get("error", {}).get("message")
-        if 400 <= response.status_code < 500 and error_message and not should_retry(response):
-            raise Exception("{} Client Error, error message: {}".format(response.status_code, error_message))
```
Comment (savan-chovatiya) [inline tap_google_analytics/client.py:261]: Add PR comment here about why the above condition is removed.

File: tap_google_analytics/client.py [hunk 2]
```diff
@@ -15,6 +15,63 @@
...
+        "raise_exception": GoogleAnalyticsInternalServerError,
+        "message": "An internal error has occurred at GoogleAnalytics's end."
+    },
+    503: {
+        "raise_exception": GoogleAnalyticsBackendError,
+        "message": "The service was unable to process the request."
+    }
+}
+
+def get_exception_for_status_code(status_code):
```
Comment (dbshah1212) [inline tap_google_analytics/client.py:71]: Do add function coments.

File: tap_google_analytics/client.py [hunk 3]
```diff
@@ -232,10 +303,9 @@ def _ensure_access_token(self):
     # Backoff Max Time: try 1 (wait 10) 2 (wait 100) 3 (wait 1000) 4
     # Gives us waits of: (10 * 10 ^ 0), (10 * 10 ^ 1), (10 * 10 ^ 2)
     #
-    # backoff for Timeout exception is already included in "requests.exceptions.RequestException"
-    # as it is the parent class of "Timeout" error
+    # added the backoff for the parent class GoogleAnalyticsClientError and for Timeout error
     @backoff.on_exception(backoff.expo,
-                          (requests.exceptions.RequestException),
+                          (requests.exceptions.Timeout, requests.exceptions.ConnectionError, GoogleAnalyticsClientError),
```
Comment (dbshah1212) [inline tap_google_analytics/client.py:308]: Do we need to retry for every error?

File: tap_google_analytics/client.py (hunk 2 above)
Comment (namrata270998) [reply tap_google_analytics/client.py:71]: Added function comments

File: tap_google_analytics/client.py (hunk 3 above)
Comment (namrata270998) [reply tap_google_analytics/client.py:309]: We are not retrying for every error, we have a giveup function which has specific codes to e retried

File: tap_google_analytics/client.py [hunk 4]
```diff
@@ -255,12 +326,8 @@ def _make_request(self, method, url, params=None, data=None):
             response = self.session.post(url, headers=headers, params=params, json=data, timeout=self.request_timeout)
         else:
             response = self.session.request(method, url, headers=headers, params=params, timeout=self.request_timeout)
-
-        error_message = _is_json(response) and response.json().get("error", {}).get("message")
```
Comment (namrata270998) [inline tap_google_analytics/client.py:259]: The removed code was throwing exceptions based on status codes and on the should_retry() method which checks if the status code is retryable, if yes raises the exception through raise_for_status(). However, we have now moved to custom exceptions, which will be raised based on corresponding status codes.

File: tap_google_analytics/client.py (hunk 1 above)
Comment (namrata270998) [reply tap_google_analytics/client.py:261]: Added PR comment

File: tests/unittests/test_client_utilities.py [hunk 5]
```diff
@@ -154,46 +162,73 @@ def test_timeout_error_get_retry(self, mocked_time_sleep, mocked_session_post, m
 
     def test_503_backend_triggers_no_retry(self, mocked_time_sleep, mocked_session_post, mocked_session_request, mocked_request_post):
         client = Client(self.config, self.config_path)
-        with self.assertRaisesRegex(requests.exceptions.RequestException, 'Raise for status'):
+        with self.assertRaises(GoogleAnalyticsClient.GoogleAnalyticsBackendError):
             client.post("503-BACKENDERROR")
 
         # Assert we gave up only after the first try
         self.assertEqual(mocked_session_post.call_count, 2)
```
Comment (dbshah1212) [inline tests/unittests/test_client_utilities.py:169]: Why does it call for 2 times instead of 1?

**Review by dbshah1212 — APPROVED** [review]

File: tests/unittests/test_client_utilities.py (hunk 5 above)
Comment (namrata270998) [reply tests/unittests/test_client_utilities.py:169]: One for the actual call and the other for post call in ensure_access_token()

**Review by KrisPersonal — COMMENTED** [review]
As per google doc, it says only certain error codes are retryable and there is a specific logic mentioned in the doc as to how many times that retries are allowed and how the logic is implemented? Can you point me to where the logic is implemented as recommended by Google Doc?

**Comment by namrata270998** [comment]
> As per google doc, it says only certain error codes are retryable and there is a specific logic mentioned in the doc as to how many times that retries are allowed and how the logic is implemented? Can you point me to where the logic is implemented as recommended by Google Doc?

Sure @KrisPersonal, It was already implemented as a part of the existing code, hence we haven't made any changes to it. Please find the existing code [here](https://github.com/singer-io/tap-google-analytics/pull/69/files#diff-6969cbfd5e0a1f5cbac30a4e1779fe7f345bb6fc8d52c4b2bafbc0f424178d9bR147) for the giveup criteria.

**Review by KrisPersonal — APPROVED** [review]

**Pull request merged**